net/connUDP.go

```diff
@@ -208,86 +207,64 @@ func (c *UDPConn) Close() error {
 	return c.connection.Close()
 }
 
-func (c *UDPConn) writeToAddr(multicastHopLimit int, iface net.Interface, srcAddr net.Addr, port string, raddr *net.UDPAddr, buffer []byte) error {
-	netType := "udp4"
-	if IsIPv6(raddr.IP) {
-		netType = "udp6"
-	}
-	addrMask := srcAddr.String()
-	addr := strings.Split(addrMask, "/")[0]
-	if strings.Contains(addr, ":") && netType == "udp4" {
-		return nil
-	}
-	if !strings.Contains(addr, ":") && netType == "udp6" {
-		return nil
-	}
+func (c *UDPConn) writeToAddr(iface *net.Interface, src *net.IP, multicastHopLimit int, raddr *net.UDPAddr, buffer []byte) error {
+	var pktSrc net.IP
 	var p packetConn
-	if netType == "udp4" {
-		p = newPacketConnIPv4(ipv4.NewPacketConn(c.connection))
-	} else {
+	if IsIPv6(raddr.IP) {
 		p = newPacketConnIPv6(ipv6.NewPacketConn(c.connection))
+		pktSrc = net.IPv6zero
+	} else {
+		p = newPacketConnIPv4(ipv4.NewPacketConn(c.connection))
+		pktSrc = net.IPv4zero
 	}
-
-	if err := p.SetMulticastInterface(&iface); err != nil {
-		return err
-	}
-	p.SetMulticastHopLimit(multicastHopLimit)
-	ip := net.ParseIP(addr)
-	if ip == nil {
-		return fmt.Errorf("cannot parse ip (%v) for iface %v", ip, iface.Name)
+	if src != nil {
+		pktSrc = *src
 	}
+
 	if c.closed.Load() {
 		return ErrConnectionIsClosed
 	}
+	if err := p.SetMulticastInterface(iface); err != nil {
+		return err
+	}
+	if err := p.SetMulticastHopLimit(multicastHopLimit); err != nil {
+		return err
+	}
 	_, err := p.WriteTo(buffer, &ControlMessage{
-		Src:     ip,
+		Src:     pktSrc,
 		IfIndex: iface.Index,
 	}, raddr)
 	return err
 }
 
-func (c *UDPConn) WriteMulticast(ctx context.Context, raddr *net.UDPAddr, hopLimit int, buffer []byte) error {
+func (c *UDPConn) WriteMulticast(ctx context.Context, raddr *net.UDPAddr, iface *net.Interface, src *net.IP, hopLimit int, buffer []byte) error {
 	if raddr == nil {
 		return fmt.Errorf("cannot write multicast with context: invalid raddr")
 	}
 	if _, ok := c.packetConn.(*packetConnIPv4); ok && IsIPv6(raddr.IP) {
 		return fmt.Errorf("cannot write multicast with context: invalid destination address")
 	}
 
-	ifaces, err := net.Interfaces()
-	if err != nil {
-		return fmt.Errorf("cannot write multicast with context: cannot get interfaces for multicast connection: %w", err)
-	}
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	for _, iface := range ifaces {
-		if iface.Flags&net.FlagMulticast == 0 {
-			continue
-		}
-		ifaceAddrs, err := iface.Addrs()
+	if iface == nil {
```

Comment:
pls, add an option set that allows sending the message without a defined interface. (it uses any interface chosen by OS)

https://github.com/plgd-dev/go-coap/blob/jkralik/feature/multicastToAnyInterface/udp/discover.go#L107

Reply:
you can just use "nil" for the interface and it automatically cycles through all available interfaces (like the code did before).
I am not sure the concept of "what the operation system chooses" really works well, you often end up using a unicast default route.

But it wouldn't be difficult to add a "WithAnyInterface()" multicast option... just needs an additional flag in the mcoption struct to differentiate between "all interfaces" and "any interface".

Reply:
`WithAnyInterface()` is needed for android 30+.  The call `net.Interfaces()` returns an error with permission denied so I want to have the option to use any interface.

Reply:
I have pushed an update, please see if this works for you (especially from the API interface).